examples/gpt-2/main.cpp

```diff
@@ -872,30 +951,34 @@ int main(int argc, char ** argv) {
         test_gpt_tokenizer(vocab, params.token_test);
     }
 
-    // keep this buffer alive while evaluating the model
-    ggml_backend_buffer_t buf_compute;
-
-    struct ggml_allocr * allocr = NULL;
-    // allocate the compute buffer
+    // create the backend scheduler
+    // the scheduler handles the allocation of the compute buffers and the scheduling of the computation between the different backends
+    ggml_backend_sched_t sched;
     {
-         // alignment required by the backend
-        size_t align = ggml_backend_get_alignment(model.backend);
-        allocr = ggml_allocr_new_measure(align);
+        // initialize the scheduler
+        sched = ggml_backend_sched_new(model.backends.data(), model.backends.size());
 
         // create the worst case graph for memory usage estimation
         int n_tokens = std::min(model.hparams.n_ctx, params.n_batch);
         int n_past = model.hparams.n_ctx - n_tokens;
-        struct ggml_cgraph * gf = gpt2_graph(model, allocr, n_past, std::vector<gpt_vocab::id>(n_tokens, 0));
+        struct ggml_cgraph * gf = gpt2_graph(model, n_past, std::vector<gpt_vocab::id>(n_tokens, 0));
 
-        // compute the required memory
-        size_t mem_size = ggml_allocr_alloc_graph(allocr, gf);
+        ggml_backend_sched_init_measure(sched, gf);
```

Comment:
This also works, correct? Just for my understanding.

```diff
diff --git a/examples/gpt-2/main.cpp b/examples/gpt-2/main.cpp
index 1a24925..a62a2b9 100644
--- a/examples/gpt-2/main.cpp
+++ b/examples/gpt-2/main.cpp
@@ -956,13 +956,13 @@ int main(int argc, char ** argv) {
     ggml_backend_sched_t sched;
     {
         // initialize the scheduler
-        sched = ggml_backend_sched_new(model.backends.data(), model.backends.size());
-
         // create the worst case graph for memory usage estimation
         int n_tokens = std::min(model.hparams.n_ctx, params.n_batch);
         int n_past = model.hparams.n_ctx - n_tokens;
         struct ggml_cgraph * gf = gpt2_graph(model, n_past, std::vector<gpt_vocab::id>(n_tokens, 0));
 
+        sched = ggml_backend_sched_new(model.backends.data(), model.backends.size());
+
         ggml_backend_sched_init_measure(sched, gf);
 
```

Reply:
Yes, that should work since it pre-allocates the inputs from a different buffer, so the `gpt2_graph` doesn't need the sched object.

However, the previous pattern of allocating the inputs manually by calling `ggml_allocr_alloc` is still supported by using `ggml_backend_sched_get_tallocr` to obtain the allocators for each backend. In that case, the sched object would need to be created before the graph.